src/liballoc/string.rs

```diff
@@ -1031,6 +1031,57 @@ impl String {
         ch
     }
 
+    /// Retains only the characters specified by the predicate.
+    ///
+    /// In other words, remove all characters `c` such that `f(c)` returns `false`.
+    /// This method operates in place and preserves the order of the retained
+    /// characters.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(string_retain)]
+    ///
+    /// let mut s = String::from("f_o_ob_ar");
+    ///
+    /// s.retain(|c| c != '_');
+    ///
+    /// assert_eq!(s, "foobar");
+    /// ```
+    #[inline]
+    #[unstable(feature = "string_retain", issue = "0")]
+    pub fn retain<F>(&mut self, mut f: F)
+        where F: FnMut(char) -> bool
```

Comment:
For `Vec` and `HashMap`, it's trivial to pass a `&mut` instead of `&`. For `String`, however, the code would have to be changed to write the decoded `char` back into the buffer. Because the new `char` could be larger than the original, it would significantly increase the complexity and perhaps require the allocation of a second buffer to manage it. I don't think that's the right direction to go in, in this case.